File: 0896-Monotonic-Array/py-0896/Solution2.py
# Source : https://leetcode.com/problems/monotonic-array/
# Author : penpenps
# Time   : 2019-07-29

from typing import List

# One pass
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def isMonotonic(self, A: List[int]) -> bool:
        n = len(A)
        trend = 0
        for i in range(1, n):
            if trend == 0:
                if A[i] > A[i-1]:
                    trend = 1
                elif A[i] < A[i-1]:
                    trend = -1
            elif trend * (A[i] - A[i-1]) < 0:
                return False
        return True

File: 0001-0500/0115-Distinct-Subsequences/java-0115/src/Solution2.java
/// Source : https://leetcode.com/problems/distinct-subsequences/description/
/// Author : liuyubobobo
/// Time   : 2018-04-23

/// Dynamic Programming
/// Time Complexity: O(s * t)
/// Space Complexity: O(t)
public class Solution2 {

    public int numDistinct(String s, String t) {

        int[] dp = new int[t.length() + 1];
        dp[0] = 1;

        for(int i = 1 ; i <= s.length() ; i ++)
            for(int j = t.length() ; j >= 1 ; j --)
                if(s.charAt(i - 1) == t.charAt(j - 1))
                    dp[j] += dp[j - 1];


        return dp[t.length()];
    }

    public static void main(String[] args) {

        String s1 = "rabbbit", t1 = "rabbit";
        System.out.println((new Solution()).numDistinct(s1, t1));
        // 3

        String s2 = "babgbag", t2 = "bag";
        System.out.println((new Solution()).numDistinct(s2, t2));
        // 5
    }
}
